Add wallet dumping and loading feature

Aschen opened the pull request:
## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts. Confirmation: ____


## What's in this pull request?

I wanted to run multiple node mining for the same wallet address so I add a feature to load a wallet from a previously dumped wallet.

The `Core` class is now construct with an options hash. For now this options hash contains only `walletSeed` which is the seed used to recover an existing wallet.  
The wallet seed is just the private key serialized and in hex format.  

```js
var walletSeed = $.wallet.dump();
var restoredWallet = await Wallet.load(walletSeed);
```

I have also modified the node and browser client accordingly.  
Node : `--wallet-seed=<wallet-seed>`
Browser : 
![nimiq_load](https://user-images.githubusercontent.com/4447392/27233646-7f724b7c-52ba-11e7-8cd0-3079517f16f4.png)

The browser client is also using the HEX version of address for transaction like in the official betanet (instead of the BASE64) .

mar-v-in commented on src/main/generic/consensus/primitive/KeyPair.js:
```diff
@@ -34,6 +38,10 @@ class KeyPair extends Primitive {
         return o instanceof KeyPair && super.equals(o);
     }
+
+    toHex() {
```
This should already by there because `KeyPair` extends `Primitive`

mar-v-in commented on src/main/generic/consensus/primitive/PublicKey.js:
```diff
@@ -28,5 +28,9 @@ class PublicKey extends Primitive {
         return new Address((await Hash.light(this.serialize())).subarray(0, 20));
     }
+
+    toHex() {
```
This should already by there because `PublicKey` extends `Primitive`

mar-v-in commented on src/main/platform/browser/Nimiq.js:
```diff
@@ -81,12 +81,12 @@ class Nimiq {
+        if (!options.path) {
             if (Nimiq._currentScript && Nimiq._currentScript.src.indexOf('/') !== -1) {
-                path = Nimiq._currentScript.src.substring(0, Nimiq._currentScript.src.lastIndexOf('/') + 1);
+                var path = Nimiq._currentScript.src.substring(0, Nimiq._currentScript.src.lastIndexOf('/') + 1);
```
Don't ever use `var` (enforced via eslint)

mar-v-in commented on clients/browser/index.html:
```diff
@@ -431,25 +442,61 @@
+    }
+
+    function createWallet() {
+      Nimiq.init(function($) {
```
Code duplication

mar-v-in replied:
`createWallet()` doesn't create a Wallet if there is already one in index db. Choose a different name.

mar-v-in replied:
Any reason you didn't cover the code duplication yet (better put the call to Nimiq.init into it's own function with options param to not have the code twice).

Aschen commented:
@mar-v-in I don't have access to the CI at url https://travis-ci.com/nimiq-network/core/jobs/79577261.
Can you give me access or tell me how I can check the indentation issues ?

mar-v-in commented:
`node_modules/.bin/gulp eslint` or just run eslint against the changed files (as there are a lot of unrelated warnings). I will check how to use travis-ci.org instead of .com for the repo, so everyone can access the reports.

Aschen commented:
It's a good idea to have an account with multi wallets, maybe if I have more time I will try to do something ! 
But for now I think dumping and loading previously created wallet is also an important thing to keep track of your Nimiq across device.

Aschen commented:
I added check for invalid wallet seed + the spec to ensure we won't have regression on this.

Aschen commented:
Any news on this PR ? @mar-v-in @paberr

paberr commented:
I'll try to take a look at it today. I'm sorry, I was quite busy the last few days.

paberr commented:
I suggest to further improve the UI a bit (as in #170) and to update the repo, since there is a conflict to be resolved. Apart from that it looks good to me. Sorry again that it took me quite long to take a look at it.

Aschen commented:
I have added the UI improvement here [6450da7](https://github.com/nimiq-network/core/pull/168/commits/6450da776dc3062f25eac83d90a848abe88db526).
No problem for the response time, it's just more and more people are asking on slack how to run multiple nimiq instance on the same wallet or how to backup/recover a wallet.
Hope your ICO will go well :-)

paberr commented:
Regarding the UI improvement, let me a bit more specific.
I checked out your repo on the master branch and looked at the UI there.

I am not sure how it is supposed to look, but in my Chrome it looks like that:
![screen shot 2017-06-26 at 10 54 27](https://user-images.githubusercontent.com/13586855/27531701-e9ba40b6-5a5d-11e7-8ffd-16e62e85257f.png)

Note the "Initializing" in the top right corner. So it seems to break the existing design at the moment. ;-)

So, my suggestions are the following:
- Drastically shorten the form field, put the whole form into the wallet box (that's why I was referring to #170).
- I think the UI (also in #170) is not yet clear enough about how to start without having a wallet seed.
It is very implicit that one can leave the text field empty and then "recover" from an empty seed.
- It would be great if the export button is only visible after one either created the new wallet or recovered from an existing seed.

Aschen commented:
Here you can see the UI : 
![nimiq_2](https://user-images.githubusercontent.com/4447392/27539116-831be1de-5a7b-11e7-9068-569320ba8ee7.png)
![nimiq_1](https://user-images.githubusercontent.com/4447392/27539117-83450f3c-5a7b-11e7-8990-3f721f1c8e7c.png)

sisou commented on dist/API_DOCUMENTATION.md:
````diff
@@ -206,24 +231,30 @@ No public properties.
 
+Dump and restore wallet:  
+```js
+var walletSeed = $.wallet.dump();
````
This function is now called `copyWalletSeed()`

Aschen replied:
I named the function `dump` because in theory it's not really a seed but only the private key.
Maybe rename all reference to a seed ? (like the variable name also..)

paberr commented on clients/browser/index.html:
```diff
@@ -27,44 +31,73 @@
+                    Load an existing browser wallet or create a new wallet:
+                    <button class="greenbutton" onclick="startNimiq()"><strong>Create/Load Wallet</strong></button>
+                </div>
+                <div id='loadWalletFromSeed'>
```
Use `"` for attributes instead of `'`. Also at some other locations in the code.

paberr commented on clients/browser/index.html:
```diff
@@ -444,28 +518,62 @@
+        mainContent.appendChild(input);
+        input.value = $.wallet.keyPair.privateKey.toHex(); //$.wallet.dump();
+        input.select();
+        document.execCommand("copy");
```
It seems like this is supposed to work in most browsers, right?
https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand#Browser_compatibility

Aschen replied:
Correct, but I don't know how far you are willing to support. I know there is an other implementation of this for older browser but its need flash ..

paberr replied:
No, we definitely don't want to have flash in there.
An option would be to display the input element with the seed on it, so if copying to clipboard fails, one can still manually copy it.

paberr commented on clients/browser/index.html:
```diff
@@ -444,28 +518,62 @@
+        input.select();
+        document.execCommand("copy");
+        mainContent.removeChild(input);
+        alert("Wallet seed copied to clipboard.");
```
This message is currently triggered even if `document.execCommand('copy')` fails.
It would be better to check whether it actually worked (not sure if that is easy).

Aschen replied:
From this gist : https://gist.github.com/jonrohan/81085b119d16cdd7868a

> Browser copy to clipboard. We use this with useragent sniffing to enable it. Because there's no way (that we know of) to feature detect clipboard behavior yet.

Aschen replied:
I added a console log of the wallet seed and modified the message accordingly

paberr commented on clients/browser/index.html:
```diff
@@ -444,28 +518,62 @@
 
-    document.getElementById('blockchain-title').classList.add('initializing');
+        var mainContent = document.getElementById('content');
+        var input = document.createElement("INPUT");
```
In JavaScript code, we use `'` instead of `"` for strings. Also occurs at some other locations.

paberr commented on clients/browser/style.css:
```diff
@@ -37,29 +48,102 @@ body {
     letter-spacing: 2px;
 }
+header a.headerlink {
+    float:right;
```
Leave a space between the CSS attribute and the value. Same holds for the whole file.

paberr commented on clients/browser/style.css:
```diff
@@ -37,29 +48,102 @@ body {
 
 header .name {
+      color:#FFF;
 	line-height: 85px;
```
Since we're modifying this file anyway, it would be great if you could make the indentation consistent. :-)

paberr commented on clients/nodejs/index.js:
```diff
@@ -22,27 +23,42 @@ if (argv['log-tag']) {
     }
     argv['log-tag'].forEach((lt) => {
-        let s = lt.split(':');
+        const s = lt.split(':');
```
:+1:

paberr commented on clients/nodejs/index.js:
```diff
@@ -22,27 +23,42 @@ if (argv['log-tag']) {
 }
 
-console.log('Nimiq NodeJS Client starting (host=' + host + ', port=' + port + ', miner=' + !!miner + ', passive=' + !!passive + ')');
+console.log(`Nimiq NodeJS Client starting (host=${host}, port=${port}, miner=${!!miner}, passive=${!!passive})`);
```
:+1:

paberr commented on src/main/generic/Core.js:
```diff
@@ -1,9 +1,9 @@
     }
 
-    async _init() {
+    async _init({ walletSeed }) {
```
To be honest, I never saw this kind of parameter list.
Do you have a reference for this JS magic?

jmingov replied:
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Search for: _Unpacking fields from objects passed as function parameter_

paberr replied:
Thanks for the reference. It looks like this is not supported by all major browsers.
So, I suggest to make the unpacking more explicit.

paberr commented on clients/nodejs/index.js:
```diff
@@ -22,27 +23,42 @@ if (argv['log-tag']) {
-(new Nimiq.Core()).then($ => {
-    console.log('Blockchain: height=' + $.blockchain.height + ', totalWork=' + $.blockchain.totalWork + ', headHash=' + $.blockchain.headHash.toBase64());
+const options = {
+    walletSeed
```
This should be set explicitly here.
Something like
```js
{
    'walletSeed': walletSeed
}
```

Aschen replied:
Should I use the more explicit syntax or use the ES6 syntax ? (I'm aware that this syntax can be destabilizing at first)

paberr replied:
Since I think we don't use shorthand property names anywhere else (I think we only use shorthand method names), I would go with the explicit syntax here.

paberr commented on src/main/generic/Core.js:
```diff
@@ -16,7 +16,11 @@ class Core {
 
         // Wallet
-        this.wallet = await Wallet.getPersistent();
+        if (walletSeed === undefined) {
```
Be careful! I'm not sure whether this works as supposed to.
[Here](https://github.com/nimiq-network/core/pull/168/files#diff-0c66691a731f5ff57c3398eaffbe3aa8R16) (`clients/nodejs/index.js`, line 16), you initialise `walletSeed` as `null` if not specified. Now you're checking against `undefined`.

Aschen replied:
Sure, thanks for that

paberr commented on yarn.lock:
```diff
@@ -5023,6 +5023,10 @@ safe-buffer@^5.0.1, safe-buffer@~5.0.1:
   version "5.0.1"
   resolved "https://registry.yarnpkg.com/safe-buffer/-/safe-buffer-5.0.1.tgz#d263ca54696cd8a306b5ca6551e92de57918fbe7"
 
+sdp@^1.5.0:
```
I'm wondering why there are two new dependencies in the `yarn.lock` file. Any ideas?

Aschen replied:
As far as I can see it's a dependency for `webrtc-adapter`. I don't remember adding a new package, just running `yarn install`

paberr replied:
I guess then it is fine.

paberr commented:
I just reviewed the current state. Most things look good to me. I'm just nit-picking regarding coding style.
Also, I have one or two questions that should be answered.

I hope the feedback helps! I have to apologise for the delay in reviewing this PR.
Thanks to @Aschen, @sisou, @happyrobotstudio for your valuable contributions! 👍

Aschen commented:
Thanks for the review @paberr :)
I forgot to tell thats @happyrobotstudio have made some UI change since last version, you can see it here : https://aschen.ovh/nimiq/
I have also add a screenshot 
![image](https://user-images.githubusercontent.com/4447392/27788953-d4cf26b6-5fea-11e7-8bca-49e1c4455bc1.png)

sisou commented:
Congratulations! =)
